openedx/core/lib/extract_archives.py

```diff
@@ -66,12 +67,51 @@ def safemembers(members, base):
                       finfo.name)
             raise SuspiciousOperation("Dev file")
 
-    return members
-
 
 def safetar_extractall(tar_file, path=".", members=None):  # pylint: disable=unused-argument
     """
-    Safe version of `tar_file.extractall()`.
+    Check if Tar file is safe and Extract
     """
     path = str(path)
-    return tar_file.extractall(path, safemembers(tar_file, path))
+    checkmembers(tar_file.getmembers(), path)
+    return tar_file.extractall(path)
+
+
+class ZipMemberAdapter:
+    """
+    Adapter ZipInfo to Member
+    """
+    def __init__(self, zipinfo):
+        self.name = zipinfo.filename
+        self.issym = lambda: False
+        self.islnk = lambda: False
+        self.isdev = lambda: False
+
+
+def safezip_extractall(zip_file, path=".", members=None):  # pylint: disable=unused-argument
+    """
+    Check if Zip file is safe and Extract
+    """
+    path = str(path)
+    checkmembers([
+        ZipMemberAdapter(zipinfo)
+        for zipinfo in zip_file.infolist()
+    ], path)
+    return zip_file.extractall(path)
+
+
+def safe_extractall(file_name, course_dir):
+    """
+    Extract Zip or Tar files
+    """
+    try:
+        if file_name.endswith('.zip'):
+            the_file = ZipFile(file_name, 'r')
+            safezip_extractall(the_file, (course_dir + '/'))
+        elif file_name.endswith('.tar.gz'):
+            the_file = tarfile.open(file_name)
+            safetar_extractall(the_file, (course_dir + '/'))
+    except Exception as e:
+        raise e
+    finally:
+        the_file.close()
```

Comment:
I refactor the code to have it's own context manager

Reply:
I think you misunderstood what I meant. Creating a context manager for this interface which is then used in only one single place makes the code harder to understand IMHO. If it was widely used it'd be different. 

What I mean was that instead of having this nested try-finally exception structure you can use the context manager that is already provided by [ZipFile](https://docs.python.org/3.10/library/zipfile.html#zipfile-objects) and [tarfile.open](https://docs.python.org/3/library/tarfile.html#examples) to avoid handling and re-raising the exception. i.e. something like this :

```python
if file_name.endswith('.zip'):
   with ZipFile(file_name, 'r') as zipfile:
        safezip_extractall(zipfile, (output_dir +'/'))
elif file_name.endswith('.tar.gz'):
   with tarfile.open(file_name) as tarfile:
        safetar_extractall(tarfile, (output_dir+'/'))
```

I don't think there is anything wrong with the code, either the previous one or this one, however I think the refactoring should make the code easier to understand. A custom context manager isn't doing that here.

Reply:
> I refactor the code to have it's own context manager

This is the approach that I proposed and implemented:

```python
def safe_extractall(file_name, output_dir):
    """
    Extract Zip or Tar files
    """
    with SafeExtractor(file_name, output_dir) as extractor:
        extractor.extract_all()
```

`SafeExtractor` is a context manager that is file agnostic.

I think it work way better this way, since I can abstract the differences on how to get the members:

```python
    @property
    def members(self):
        """
        Get members from archive
        """
        if self.file_name.endswith('.zip'):
            return [
                ZipMemberAdapter(zipinfo)
                for zipinfo in self.the_file.infolist()
            ]
        return self.the_file.getmembers()
```

And keep the complexity in this code.

I actually got this approach from the great presentation [Raymond Hettinger - Beyond PEP 8 -- Best practices for beautiful intelligible code](https://www.youtube.com/watch?v=wf-BqAjZb8M)

Reply:
done